components/script/dom/document.rs

```diff
@@ -829,27 +831,37 @@ impl Document {
             // Update the focus state for all elements in the focus chain.
             // https://html.spec.whatwg.org/multipage/#focus-chain
             if focus_type == FocusType::Element {
-                self.send_to_constellation(ScriptMsg::Focus);
+                self.window().send_to_constellation(ScriptMsg::Focus);
             }
 
             // Notify the embedder to display an input method.
             if let Some(kind) = elem.input_method_type() {
-                self.send_to_constellation(ScriptMsg::ShowIME(kind));
+                let top_level_browsing_context_id = self.window().top_level_browsing_context_id();
+                self.send_to_embedder(EmbedderMsg::ShowIME(top_level_browsing_context_id, kind));
             }
         }
     }
 
     /// Handles any updates when the document's title has changed.
     pub fn title_changed(&self) {
         if self.browsing_context().is_some() {
-            self.send_title_to_constellation();
+            self.send_title_to_embedder();
         }
     }
 
     /// Sends this document's title to the constellation.
-    pub fn send_title_to_constellation(&self) {
-        let title = Some(String::from(self.Title()));
-        self.send_to_constellation(ScriptMsg::SetTitle(title));
+    pub fn send_title_to_embedder(&self) {
+        let window = self.window();
+        if window.is_top_level() {
```

Comment:
I think this is right. There is no way for a window to be top level and not be the top level browsing context, right? (that was not always the case)

Reply:
yes I think that's right, yet `window.is_top_level` checks for 'top levelness', while `window.top_level_browsing_context_id()` will always return the `top_level_browsing_context_id` regardless of whether the window on which it is called is actuall top level or not...